File: credit.c
// This program implements Luhn's algorithm to check a given credit card number and output if it's AMEX, VISA, MasterCard,
// or, INVALID.

#include <stdio.h>
#include <cs50.h>

int main(void)
{
    // Prompt user for credit card number
    long long ccnumber;
    do
    {
        ccnumber = get_long_long("Enter a card number: ");
    }
    while (ccnumber <= 0);

    // Get the length of the number (count). Initialize count and a temp. variable n, required for the calculation
    int count = 0;
    long long n = ccnumber;
    while (n != 0)
    {
        n /= 10;
        count++;
    }

    // Get the digits and store them in an array (digs) for easy retrieval. Initialize digs and m (for calc. only) before the loop.
    // Start the loop from the last digit and run it backward
    int digs[count];
    long long m = ccnumber;
    for (int i = count - 1; i >= 0; i--)
    {
        if (i == count - 1)
        {
            digs[i] = m % 10;
        }
        else
        {
            // Use m to a calculation to get successive digits
            m /= 10;
            digs[i] = m % 10;
        }
    }

    // Calculate the sum of products of 2 and every alternate number starting from last but one digit (sum1). Initialize sum1.
    int sum1 = 0;
    for (int i = count - 2; i >= 0; i = i - 2)
    {
        // If the product >= 10, split the digits and add them individually. Use n, like we did before. Intialize it.
        n = digs[i] * 2;
        if (digs[i] * 2 >= 10)
        {
            sum1 += n % 10;
            n /= 10;
            sum1 += n;
        }
        else
        {
            // If product is < 10, add it to sum1
            sum1 += digs[i] * 2;
        }
    }

    // Calculate the sum of every alternate number that wasn't chosen in the above step (sum2). Initialize sum2.
    int sum2 = 0;
    for (int i = count - 1; i >= 0; i = i - 2)
    {
        sum2 += digs[i];
    }

    // Check if the card belongs to AMEX, VISA, MasterCard, or if it's invalid
    if ((sum1 + sum2) % 10 != 0)
    {
        printf("INVALID\n");
    }
    else
    {
        // VISA has 16 digits and starts with 4
        if (digs[0] == 4 && (count == 13 || count == 16))
        {
            printf("VISA\n");
        }
        //MasterCard has 16 digits ansd starts with 51, 52, 53, 54, or 55
        else if (digs[0] == 5 && ((digs[1] == 1) || (digs[1] == 2) || (digs[1] == 3) || (digs[1] == 4) || (digs[1] == 5)) && count == 16)
        {
            printf("MASTERCARD\n");
        }
        // AMEX has 13 digits and starts with 35 or 37
        else if (digs[0] == 3 && ((digs[1] == 5) || (digs[1] == 7)) && count == 15)
        {
            printf("AMEX\n");
        }
        else
        {
            // Print invalid if all 3 checks fail
            printf("INVALID\n");
        }
    }
}

File: sentimental/credit.py
# credit.py

# Include everything in the main function to match C


def main():
    # Prompt user for credit card number
    ccnumber = int(input("Enter a card number: "))

    # Check if the number > 0
    while ccnumber <= 0:
        ccnumber = int(input("Invalid number! Enter a card number: "))

    # Get the length of the card number
    count = len(str(ccnumber))

    # Store the digits of the card in a list for convenience
    digs = []
    m = ccnumber
    for i in range(count, 0, -1):
        # Do number mod 10 to get the last digit
        if i == count:
            digs.append(m % 10)
        else:
            # Divide (floor) number by 10 and then do modulo to get subsequent digits
            m //= 10
            digs.append(m % 10)

    # Multiply digits starting from the last but one'th position by 2 and and store in sum
    sum1 = 0
    for i in range(1, count, 2):
        # If the product > 10, split the digits and add them individually to sum1
        n = digs[i] * 2
        if n >= 10:
            sum1 += n % 10
            n //= 10
            sum1 += n
        else:
            sum1 += n

    # Sum over all the remaining digits and store them in sum2
    sum2 = 0
    for i in range(0, count, 2):
        sum2 += digs[i]

    # Check if card is valid
    if (sum1 + sum2) % 10 != 0:

        print("INVALID")

    else:
        if digs[count - 1] == 4 and (count == 13 or count == 16):

            print("VISA")

        elif digs[count - 1] == 5 and (digs[count - 2] == 1 or digs[count - 2] == 2 or
                                       digs[count - 2] == 3 or digs[count - 2] == 4 or digs[count - 2] == 5) and count == 16:

            print("MASTERCARD")

        elif digs[count - 1] == 3 and (digs[count - 2] == 5 or digs[count - 2] == 7) and count == 15:

            print("AMEX")

        else:

            print("INVALID")

    # Return 0
    return 0


# Call main by default
if __name__ == "__main__":
    main()
